### PR TITLE
[refactor]: Refactor and add tests for topology

### PR DESCRIPTION
**Pull request by Erigara**
### Linked issue

Closes #{issue_number} 

### Benefits

Less code and new tests.

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [x] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: core/src/sumeragi/network_topology.rs [hunk 1]
```diff
@@ -128,47 +130,23 @@ impl Topology {
...
-        while i < self.sorted_peers.len() {
-            if to_lift_up.contains(&self.sorted_peers[i].public_key) {
-                i += 1;
-            } else {
-                observing.insert(0, self.sorted_peers.remove(i)); // This has to be insert(0) and not push in order to preserve order.
-            }
-        }
-        self.sorted_peers.extend(observing);
+        self.sorted_peers
+            .sort_by_cached_key(|peer| !to_lift_up.contains(&peer.public_key));
```
Comment (mversic) [inline core/src/sumeragi/network_topology.rs:149]: previous leader always becomes the tail

**Review by SamHSmith — APPROVED** [review]
This is good.

File: core/src/sumeragi/network_topology.rs (hunk 1 above)
Comment (SamHSmith) [reply core/src/sumeragi/network_topology.rs:149]: The test lift_up_peers demonstrates that the behavior is correct.

Comment (Erigara) [reply core/src/sumeragi/network_topology.rs:149]: you mean proxy tail? or last element in topology?

**Comment by Erigara** [comment]
Soft-fork test stop working, need to figure out what happened.

File: core/src/sumeragi/network_topology.rs (hunk 1 above)
Comment (mversic) [reply core/src/sumeragi/network_topology.rs:149]: sorry, I commented in the wrong place

File: core/src/sumeragi/network_topology.rs [hunk 2]
```diff
@@ -128,47 +130,23 @@ impl Topology {
...
     }
     /// Re-arrange the set of peers after each successful block commit.
     pub fn rotate_set_a(&mut self) {
-        let top = self.sorted_peers.remove(0);
-        self.sorted_peers.insert(
-            self.min_votes_for_commit().min(self.sorted_peers.len()),
-            top,
-        );
+        let rotate_at = self.min_votes_for_commit().min(self.sorted_peers.len());
+        self.sorted_peers[..rotate_at].rotate_left(1);
```
Comment (mversic) [inline core/src/sumeragi/network_topology.rs:144]: previous leader always becomes the tail

**Review by mversic — APPROVED** [review]

File: core/src/sumeragi/network_topology.rs (hunk 2 above)
Comment (Erigara) [reply core/src/sumeragi/network_topology.rs:144]: it's the same behavior we have before that, isn't it?

Comment (mversic) [reply core/src/sumeragi/network_topology.rs:144]: yes, it's just something I'm bringing attention to

**Comment by Erigara** [comment]
> Soft-fork test stop working, need to figure out what happened.

Aha, previous implementation of topology was buggy. 

And that's why soft-fork test wasn't testing anything.
Faulty-peer never became proxy-tail.

@SamHSmith, @mversic so i guess soft-fork should be fixed first, what do you think?

**Review by mversic — APPROVED** [review]

File: core/src/sumeragi/network_topology.rs (hunk 2 above)
Comment (SamHSmith) [reply core/src/sumeragi/network_topology.rs:144]: > previous leader always becomes the tail. @SamHSmith do you see any problem with that?

No problem. That is the intented effect. I think I left some comment somewhere in the code that says this. Although probably cryptically.

**Pull request merged**